File: fixtures/coding/batch-ingestion-go/transform.go
package pipeline

// Record transformation — sign your implementation here.
//
// All function signatures are fixed. Fill the function bodies.

// NormalizePayload normalises payload keys to snake_case.
//
// For every key in Payload:
//   - Convert PascalCase/CamelCase keys to snake_case.
//   - Trim whitespace from string values.
//   - Leave non-string values unchanged.
//
// Return a new Record with the transformed payload.
// Do NOT mutate the input record, payload, or tags.
func NormalizePayload(record Record) Record {
	return Record{}
}

// FlattenPayload flattens one level of nested map values in Payload.
//
// e.g. {"user": {"name": "alice", "id": 7}} becomes
// {"user_name": "alice", "user_id": 7, ...other keys}
//
// Only flatten one level. Non-map values stay as-is.
// Return a new Record with the flattened payload.
func FlattenPayload(record Record) Record {
	return Record{}
}

// TransformBatch applies NormalizePayload then FlattenPayload to every record.
//
// Return new records in the original order.
func TransformBatch(records []Record) []Record {
	return nil
}

File: fixtures/coding/batch-ingestion-rust/src/transform.rs
use crate::types::{Payload, Record};

/// Record transformation — sign your implementation here.
///
/// All function signatures are fixed. Fill the function bodies.

/// Normalize payload keys to snake_case and trim string payload values.
///
/// Return a new record. Do NOT mutate the input record, payload, or tags.
pub fn normalize_payload(_record: &Record) -> Record {
    Record {
        record_type: String::new(),
        ts: 0,
        payload: Payload::new(),
        tags: Default::default(),
    }
}

/// Flatten one level of nested object payload values.
///
/// `{ "user": Object({"name": "alice"}) }` becomes `{ "user_name": "alice" }`.
/// Only flatten one level. Non-object values stay as-is.
pub fn flatten_payload(_record: &Record) -> Record {
    Record {
        record_type: String::new(),
        ts: 0,
        payload: Payload::new(),
        tags: Default::default(),
    }
}

/// Apply `normalize_payload` and then `flatten_payload` to every record.
///
/// Preserve input order.
pub fn transform_batch(_records: &[Record]) -> Vec<Record> {
    Vec::new()
}
